### PR TITLE
Read vmfb files in 2GB chunks to resolve fread issues on Android

### PR DESCRIPTION
**Pull request by suryajasper**
Solves #14571

### CODE REVIEW

**Comment by suryajasper** [comment]
@benvanik can you take a look at this when you get the chance? I don't think I can assign reviewers :)

File: runtime/src/iree/base/internal/file_io.c [hunk 1]
```diff
@@ -164,12 +164,20 @@ static iree_status_t iree_file_preload_contents_impl(
...
 
-  // Attempt to read the file into memory.
-  if (fread(contents->buffer.data, 1, file_size, file) != file_size) {
-    iree_allocator_free(allocator, contents);
-    return iree_make_status(IREE_STATUS_PERMISSION_DENIED,
-                            "unable to read entire %" PRIhsz " file bytes",
-                            file_size);
+  // Attempt to read the file into memory, chunking into ~2GB segments.
+  iree_host_size_t bytes_read = 0;
+  while ( bytes_read < file_size ) {
```
Comment (benvanik) [inline runtime/src/iree/base/internal/file_io.c]: you'll need to appease the clang-format gods
```suggestion
  while (bytes_read < file_size) {
```

File: runtime/src/iree/base/internal/file_io.c [hunk 2]
```diff
@@ -164,12 +164,20 @@ static iree_status_t iree_file_preload_contents_impl(
...
+    iree_host_size_t chunk_size = file_size - bytes_read;
+    if (chunk_size > INT_MAX)
+      chunk_size = INT_MAX;
+
```
Comment (benvanik) [inline runtime/src/iree/base/internal/file_io.c]:
```suggestion
    iree_host_size_t chunk_size = iree_min(file_size - bytes_read, INT_MAX);
```

**Review by benvanik — CHANGES_REQUESTED** [review]

**Review by benvanik — APPROVED** [review]

**Review by ScottTodd — APPROVED** [review]

**Comment by ScottTodd** [comment]
GtHub's status checks seem to be stuck on this PR. Maybe push an empty commit to re-trigger?

**Comment by ScottTodd** [comment]
Weiiiird... I tried running Android benchmarks and those are also broken? https://github.com/openxla/iree/actions/runs/5823178023/job/15789460681?pr=14617 (unrelated to these code changes). If it's not a GitHub issue, I wonder if it's related to the PR coming from a fork of SHARK-Runtime (itself a fork of IREE) and not IREE itself...? Idk...

**Comment by suryajasper** [comment]
hmm, yeah it might be because the branch is based off of SHARK-Runtime. I can try making a new branch based off IREE/main and opening a new PR from there if SHARK-runtime is causing tests to fail.

**Comment by ScottTodd** [comment]
Pushing an empty commit (`git commit -m "Empty commit to poke CI." --allow-empty`) would be the simplest way to check if GitHub is just confused. I don't see a way to retrigger the checks from the UI, and I don't have permission to push to that branch myself.

LMK if I can help.

File: runtime/src/iree/base/internal/file_io.c [hunk 3]
```diff
@@ -164,12 +164,17 @@ static iree_status_t iree_file_preload_contents_impl(
...
-  if (fread(contents->buffer.data, 1, file_size, file) != file_size) {
-    iree_allocator_free(allocator, contents);
-    return iree_make_status(IREE_STATUS_PERMISSION_DENIED,
-                            "unable to read entire %" PRIhsz " file bytes",
-                            file_size);
+  // Attempt to read the file into memory, chunking into ~2GB segments.
+  iree_host_size_t bytes_read = 0;
+  while (bytes_read < file_size) {
+    iree_host_size_t chunk_size = iree_min(file_size - bytes_read, INT_MAX);
+    if (fread(contents->buffer.data + bytes_read, 1, chunk_size, file) != chunk_size) {
```
Comment (ScottTodd) [inline runtime/src/iree/base/internal/file_io.c]:
```suggestion
    if (fread(contents->buffer.data + bytes_read, 1, chunk_size, file) !=
        chunk_size) {
```

**Review by ScottTodd — APPROVED** [review]
Phew, CI is running. Looks like clang-format is still unhappy, but the rest should be good.

**Comment by benvanik** [comment]
Bots are looking happy, thanks for the fix and helping to narrow it down @suryajasper!

**Pull request merged**